### PR TITLE
[Windows] Fix Fasthash for Movies Scan

### PR DESCRIPTION
**Pull request by CrystalP**
## Description

The library scanner tries to detect unchanged directories containing movies and music videos with a fast hash algorithm, which uses XFILE::CFile::Stat() on directories.

However Stat() fails on Windows platform with local and mapped/smb drives because the paths end with a slash.

The PR removes the trailing \ and adds Doxygen documentation of the possibility of filenames with ending slash for directories.

The speedup of scan with fast hash vs regular hash varies but network filesystems have more potential for improvement.
Testing showed 15% scan time improvement on local drive and 40% for same library on a mapped network drive.

Other filesystems such as nfs have their own path sanitization and already use fasthash.
Others may not (not available to me for testing) and would need a similar change to take advantage of fasthash.

Fix not made generic (removing the ending slash before calling Stat) because that ending slash is the only hint that it's a directory and not a file. Some filesystems may need different ways to stat files and directories (for example ftp, with file listings - although broken now)

## Motivation and context

Found out while stepping through the video scanner that fast hash always fails and regular hash is used with Windows.

## How has this been tested?

Scanned movies on local drive and NAS mapped network drive.
Tried stat'ing a drive like X: or the root of a network drive \\server\drive or smb://server/drive. Those fail by design, no change.
Confirmed that the hashes were replaced with fast hashes in the db on first scan and then remained stable,
The log indicates that fast has is used.

## What is the effect on users?

Faster scan of unchanged movies and music videos with Windows.
First scan take longer to convert from regular to fast hash.

## Screenshots (if appropriate):

on a test library of 200 movies:

before
```
2024-12-14 20:56:15.754 T:21680   debug <general>: VideoInfoScanner: Skipping dir 'Y:\Movies\2012 (2009)\' due to no change
2024-12-14 20:56:15.781 T:21680   debug <general>: VideoInfoScanner: Skipping dir 'Y:\Movies\3000 Miles to Graceland (2001)\' due to no change
2024-12-14 20:56:15.784 T:21680    info <general>: VideoInfoScanner: Finished scan. Scanning for video info took 4464 ms
```

after
```
2024-12-14 20:30:37.641 T:11152   debug <general>: VideoInfoScanner: Skipping dir 'Y:\Movies\2012 (2009)\' due to no change (fasthash)
2024-12-14 20:30:37.651 T:11152   debug <general>: VideoInfoScanner: Skipping dir 'Y:\Movies\3000 Miles to Graceland (2001)\' due to no change (fasthash)
2024-12-14 20:30:37.654 T:11152    info <general>: VideoInfoScanner: Finished scan. Scanning for video info took 3150 ms
```

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Review by sundermann — APPROVED** [review]

**Comment by CrystalP** [comment]
I'm having second thoughts, can all filesystems be assumed to be able to stat files and directories the same way?
Stat has a string parameter and with the ending / removed it's not possible to know if the last component is supposed to be a file or a directory.

For example ftp, implemented with curl. Though stat'ing directories is broken today even with an ending / in path, it will be more difficult to improve if it cannot make a distinction between files and directories.

**Comment by CrystalP** [comment]
Sorry @sundermann changed my mind and made this Windows specific, in order to not impede filesystems that may need to know the difference between file and directory or that I don't have access to for testing.

PosixFile is a prime candidate for a similar change from what I read.

Another way would be to create a way for filesystems to declare capabilities but that gets complicated fast.

**Comment by CrystalP** [comment]
Jenkins build this please

File: xbmc/platform/win32/filesystem/Win32File.cpp [hunk 1]
```diff
@@ -466,7 +466,15 @@ int CWin32File::Stat(const CURL& url, struct __stat64* statData)
     return -1;
   }
 
-  if (pathnameW.length() <= 6) // 6 is length of "\\?\x:"
+  size_t pathLength{pathnameW.length()};
+  // stat of directories requires the removal of the trailing slash
+  if (pathnameW[pathLength - 1] == L'\\')
+  {
+    pathnameW.erase(pathLength - 1);
+    --pathLength;
```
Comment (thexai) [inline xbmc/platform/win32/filesystem/Win32File.cpp:474]: `--pathLength` it's a bit ugly, maybe remove

and use `pathnameW.length()` in L477 and then `pathLength` declaration can be `const size_t`

Comment (thexai) [reply xbmc/platform/win32/filesystem/Win32File.cpp:474]: Or even better:

```c++
  const size_t last{pathnameW.length() - 1};
  // stat of directories requires the removal of the trailing slash
  if (pathnameW[last] == L'\\')
    pathnameW.erase(last);
```

Comment (CrystalP) [reply xbmc/platform/win32/filesystem/Win32File.cpp:474]: That's what I originally had, and then noticed there was another use of the length in the function. (L477).
Seemed unnecessary to throw away the length and calculate it again. But I don't see a prettier way.

There is this, prettier but maybe less efficient:
```c++
  // stat of directories requires the removal of the trailing slash
  if (pathnameW.back() == L'\\')
    pathnameW.pop_back();

  if (pathnameW.length() <= 6) // 6 is length of "\\?\x:"
```

**Review by thexai — APPROVED** [review]

**Comment by CrystalP** [comment]
Missed the replacement of one @return with \return in the Doxygen info per coding guidelines. Will merge once green.

**Pull request merged**